Description: CA-81375: xe-edit-bootloader uses kpartx
Origin: upstream, commit a8bd0f84284c31bddca8d4b554ccd4a4072d624f
Author: Mike McClurg <mike.mcclurg@citrix.com>
Bug-Debian: http://bugs.debian.org/675050, http://bugs.debian.org/675055
Last-Update: Wed May 30 13:26:14 2012 +0100

--- a/scripts/xe-edit-bootloader
+++ b/scripts/xe-edit-bootloader
@@ -11,7 +11,7 @@
 
 . @INVENTORY@
 
-XE="@BINDIR@/xe"
+XE="/usr/bin/xe"
 
 device_number=
 grub_conf=
@@ -84,6 +84,8 @@
       rmdir ${mnt}
    fi
 
+   kpartx -d ${device}
+
    if [ ! -z "${vbd_uuid}" ]; then
       echo -n "Unplugging VBD: "
       ${XE} vbd-unplug uuid=${vbd_uuid} timeout=20
@@ -185,12 +187,15 @@
   exit 1
 fi
 
-echo -n "Waiting for ${device}${device_number}: "
+kpartx -av ${device}
+mapped_device="/dev/mapper/${vdi_uuid}"
+
+echo -n "Waiting for ${mapped_device}${device_number}: "
 found_device=0
 
 for ((i=0; i<5; i++)); do
    echo -n '.'
-   if [ -b ${device}${device_number} ]; then
+   if [ -b ${mapped_device}${device_number} ]; then
       found_device=1
       echo ' done'
       break
@@ -199,7 +204,7 @@
 done
 
 if [ ${found_device} -eq 0 ]; then
-   echo Device ${device}${device_number} not found.
+   echo Device ${mapped_device}${device_number} not found.
    echo You must specify the correct partition number with -p
    cleanup
    exit 1
@@ -209,13 +214,13 @@
 mnt=/var/run/fix-grub-${vdi_uuid}
 mkdir -p ${mnt}
 
-mount ${device}${device_number} ${mnt} > /dev/null 2>&1
+mount ${mapped_device}${device_number} ${mnt} > /dev/null 2>&1
 
 if [ $? -ne 0 ]; then
   echo " failed"
   echo Partitions in the VDI are:
   echo
-  ls -la ${device}*
+  ls -la ${mapped_device}*
   echo
   echo You can use the -p option to specify a partition number to mount.
   cleanup
